```bash
#!/bin/bash
#The function of this script is to remove stop codons from coding sequences and then translate them into amino acid sequences
# Define the base directory
base_dir=~/data/test/75/75

# List all files from the current directory and read each file name as the id variable
ls | while read id
do
    # Enter the ${base_dir}/${id} directory, where ${id} is the file name read above
    cd ${base_dir}/${id}
    # List all files ending with .fasta and remove the suffix, read each gene name as the gene variable
    ls *.fasta | cut -f 1 -d "." | while read gene
    do
        # Use the perl command to merge multiple lines of sequences in the ${gene}.fasta file into a single line sequence and output it to the ${gene}.tmp.fasta file
        perl -pe '$. > 1 and /^>/ ? print "\n" : chomp' ${gene}.fasta > ${gene}.tmp.fasta
        # Print a message indicating that ${gene}.fasta has been converted to a single line sequence format
        echo "${id}-${gene} has been a Single Line Fasta"
        # Use the sed command to remove the stop codons (TGA,TAA,TAG) at the end of the ${gene}.tmp.fasta file and output it to the ${id}_${gene}_no_stop_codon.fasta file
        sed -e "s/TGA$//" -e "s/TAA$//" -e "s/TAG$//" ${gene}.tmp.fasta > ${id}_${gene}_no_stop_codon.fasta
        # Print a message indicating that the stop codon has been removed
        echo "${id}-${gene} has remove the stop codon"
        # Use the transeq command to translate the nucleotide sequence in the ${id}_${gene}_no_stop_codon.fasta file into an amino acid sequence and output it to the ${id}_${gene}_protein.fasta file
        transeq -sequence ${id}_${gene}_no_stop_codon.fasta -outseq ${id}_${gene}_protein.fasta
        # Print a message indicating that it has been translated into an amino acid sequence
        echo "${id}-${gene} has trans to protein sequence"
        # Use sed command to remove extra _1 characters from ${id}_${gene}_protein.fasta file (this is generated by transeq command)
        sed -i "s/\_1//g" ${id}_${gene}_protein.fasta 
    done # End inner loop (operate on each gene)
done # End outer loop (operate on each file)
```
